### PR TITLE
Llama checkpoint

### LINKED ISSUE #116: [Super AI] Implement Gradient Checkpointing
Description:
Gradient Checkpointing is one the best techniques to reduce model memory usage when training. In Addtiion, there is notion that gradient checkpointing should only be done to attention calculation NOT the entire model https://arxiv.org/abs/2205.05198

Instructions:
- Create Option to do gradient checkpointing for the entire model
- Create Optiion to do gradient checkpointing for only self-attetion

Deliverables:
- Run on Huaweii Cloud
- PR

### PR DESCRIPTION
**Pull request by MoosaTae**
## Why this PR
Implement Gradient Checkpointing for Lightning Fabric training pipeline

## Changes
- add Gradient Checkpointing option for LLama
### To test gradient checkpointing
```
!torchrun --standalone --nproc_per_node=1 src/model/scripts/lighting_training/train.py --model_name llama --vocab_size 32000 --checkpoint 1
```
- 1 is for model 
- 2 is for self attentions only
## Related Issues
Close #116 

## Checklist
- [ ] PR should be in the [Naming convention](../../docs/PR_NAMING.md)
- [x] Assign yourself in to Assigneees
- [x] Tag related issues
- [x] Constants name should be ALL_CAPITAL, function name should be snake_case, and class name should be CamelCase
- [x] complex function/algorithm should have [Docstring](https://peps.python.org/pep-0257/)
- [x] 1 PR should not have more than 200 lines changes (Exception for test files). If more than that please open multiple PRs
- [x] At least PR reviewer must come from the task's team (model, eval, data)

### CODE REVIEW

File: src/model/scripts/lighting_training/train.py [hunk 1]
```diff
@@ -59,9 +59,9 @@
     )
     parser.add_argument(
         "--checkpoint",
-        type=bool,
-        default=False,
-        help="gradient checkpoint only available for GPTJ",
+        type=int,
+        default=0,
+        help="0 | 1 (model) | 2 (self-attentions only)",
```
Comment (boat1603) [inline src/model/scripts/lighting_training/train.py:64]: I think we should have 2 parameters. 1 is for `use_checkpointing` and 1 is for `checkpointing_position`. Is it better than 0, 1, 2?

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+            outputs += (self_attn_weights,)  # type: ignore
+
+        if use_cache:
+            outputs += (present_key_value,)  # type: ignore
+
+        return outputs  # type: ignore
+
+
+# create new Llama that call LlamaClass which edit LlamaDecoderLayerWithCheckpointing
+class LlamaModify(LlamaPreTrainedModel):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:315]: What are you modified? I think you should use the better name.

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+from torch.nn import CrossEntropyLoss
+
+from typing import Optional, Tuple, List, Union
+from transformers.modeling_outputs import (
+    BaseModelOutputWithPast,
+    CausalLMOutputWithPast,
+)
+
+
+class LlamaModel(LlamaPreTrainedModel):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:25]: How different between `LlamaModel` in this code and from `huggingface trasformers`? If you modified some function, please use class inherit.

File: src/model/scripts/lighting_training/train.py (hunk 1 above)
Comment (MoosaTae) [reply src/model/scripts/lighting_training/train.py]: done

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py (hunk 2 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/models/llama_hf/model.py]: It call LlamaModel() with have implement gradient checkpoint, rename was done.

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py (hunk 3 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:25]: I modified init to use new self.layers -> LlamaDecoderLayerWithCheckpointing(config) and forward to pass gradient_checkpointing parameters.

File: src/model/openthaigpt_pretraining_model/lightning/utils.py [hunk 4]
```diff
@@ -106,17 +108,19 @@ def __init__(
             precision=precision,
         )
         self.fabric.launch()
+        if checkpoint:
+            if checkpointing_position:
+                checkpointing = 2
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/lightning/utils.py:113]: Can you use Enum instead

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+            outputs += (self_attn_weights,)  # type: ignore
+
+        if use_cache:
+            outputs += (present_key_value,)  # type: ignore
+
+        return outputs  # type: ignore
+
+
+# create new Llama that call LlamaClass which edit LlamaDecoderLayerWithCheckpointing
+class LlamaNewCheckpoint(LlamaPreTrainedModel):
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:315]: Then you can just inherit class LlamaForCausalLMWithCheckpoint(LlamaForCausalLM) like this and just modify few functions

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+    cfg = LlamaConfig(
+        vocab_size=vocab_size,
+        hidden_size=1024,
+        num_hidden_layers=8,
+        num_attention_heads=8,
+        hidden_act="silu",
+        max_position_embeddings=context_length,
+    )
+    if use_checkpointing == 2:
+        model = LlamaNewCheckpoint(cfg)
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:494]: Should be name LlamaForCausalLMNewCheckpoint

**Comment by new5558** [comment]
Can you run it on Lanta Slurm cluster and see if it works?

File: src/model/scripts/lighting_training/train.py [hunk 7]
```diff
@@ -61,7 +61,13 @@
...
         type=bool,
         default=False,
-        help="gradient checkpoint only available for GPTJ",
+        help="gradient checkpointing",
+    )
+    parser.add_argument(
+        "--check_position",
+        type=bool,
+        default=False,
+        help="False (model) | True (self-attentions only)",
```
Comment (boat1603) [inline src/model/scripts/lighting_training/train.py:70]: I think this line is too weird. Is it better if we use `--checkpoint_only_attention`?

File: src/model/openthaigpt_pretraining_model/lightning/utils.py (hunk 4 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py]: done

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py [hunk 8]
```diff
@@ -0,0 +1,508 @@
...
+def make_model_llama(vocab_size, context_length, use_checkpointing):
+    cfg = LlamaConfig(
+        vocab_size=vocab_size,
+        hidden_size=1024,
+        num_hidden_layers=8,
+        num_attention_heads=8,
+        hidden_act="silu",
+        max_position_embeddings=context_length,
+    )
+    if use_checkpointing == 2:
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:493]: I think you should feed `use_checkpointing` and `checkpoint_only_attention` in the input of model. You use `use_checkpointing==2` we don't know how different between `0`, `1`, and `2`.

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py (hunk 3 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/models/llama_hf/model.py]: done

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py (hunk 6 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/models/llama_hf/model.py]: done

File: src/model/scripts/lighting_training/train.py (hunk 7 above)
Comment (MoosaTae) [reply src/model/scripts/lighting_training/train.py:70]: done

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py (hunk 8 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:387]: I change It in utils.py now to use enum instead, Must I do this?

Comment (boat1603) [reply src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:387]: For me I will write
```python
if use_checkpointing and checkpoint_only_attention:
    model = LlamaForCausalLMNewCheckpoint(cfg)
    print("use gradient checkpointing only attentions")
else:
    model = LlamaForCausalLM(cfg)
```

Is it clearer than `use_checkpointing=2`?
You can pass 2 parameters without using Enum and some login in above.

Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/models/llama_hf/model.py]: good Idea, done.

File: src/model/openthaigpt_pretraining_model/lightning/utils.py [hunk 9]
```diff
@@ -106,17 +108,16 @@ def __init__(
...
+            self.model = model = make_model_llama(
                 vocab_size=vocab_size,
-                hidden_size=1024,
-                num_hidden_layers=8,
-                num_attention_heads=8,
-                hidden_act="silu",
-                max_position_embeddings=context_length,
+                context_length=context_length,
+                use_checkpointing=checkpoint,
+                checkpoint_only_attentions=checkpoint_only_attention,
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/lightning/utils.py:118]: `checkpoint_only_attentions` or `checkpoint_only_attention` you should use one.

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+            logits=logits,
+            past_key_values=outputs.past_key_values,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
+
+
+def make_model_llama(
+    vocab_size, context_length, use_checkpointing, checkpoint_only_attentions
+):
```
Comment (boat1603) [inline src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:380]: We shold have argument description and hint. You can see an example [here](https://github.com/OpenThaiGPT/Pretraining-ELECTRA/blob/36acaf52364c4159adfdb2a5e975bd2600f92315/src/electra_pretraining/cli_tools/pretraining.py#L157).

File: src/model/openthaigpt_pretraining_model/lightning/utils.py (hunk 9 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py]: done

File: src/model/openthaigpt_pretraining_model/models/llama_hf/model.py (hunk 10 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/models/llama_hf/model.py:383]: done

**Review by boat1603 — APPROVED** [review]

File: src/model/openthaigpt_pretraining_model/lightning/utils.py [hunk 11]
```diff
@@ -106,17 +108,16 @@ def __init__(
             precision=precision,
         )
         self.fabric.launch()
+
         if model_name == "llama":
             model_name = LLAMA_MODEL  # for tokenizer
-            cfg = LlamaConfig(
+            self.model = model = make_model_llama(
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/lightning/utils.py:114]: why ` self.model = model = make_model_llama(`

File: src/model/openthaigpt_pretraining_model/lightning/utils.py [hunk 12]
```diff
@@ -106,17 +108,16 @@ def __init__(
...
-                max_position_embeddings=context_length,
+                context_length=context_length,
+                use_checkpointing=checkpoint,
+                checkpoint_only_attention=checkpoint_only_attention,
             )
-            self.model = model = LlamaForCausalLM(cfg)
+
         elif model_name == "gptj":
             model_name = GPTJ_MODEL  # for tokenizer
             self.model = model = make_model_gptj(
```
Comment (new5558) [inline src/model/openthaigpt_pretraining_model/lightning/utils.py:123]: why `  self.model = model = make_model_gptj(`

**Review by new5558 — APPROVED** [review]
Fix this and LGTM

File: src/model/openthaigpt_pretraining_model/lightning/utils.py (hunk 11 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:114]: done

File: src/model/openthaigpt_pretraining_model/lightning/utils.py (hunk 12 above)
Comment (MoosaTae) [reply src/model/openthaigpt_pretraining_model/lightning/utils.py:123]: done
